src/components/labeledInput.tsx

```diff
@@ -0,0 +1,229 @@
+import React, { ReactElement } from 'react'
+import { TextField, Box, Checkbox, FormControlLabel } from '@material-ui/core'
+import { Control, Controller, RegisterOptions } from 'react-hook-form'
+import InputMask from 'react-input-mask'
+import locationPostalCodes from '../data/locationPostalCodes'
+import { PatternConfig } from './Patterns'
+
+interface Errors {
+  [key: string]: {type: string}
+}
+
+type Register = (rules?: RegisterOptions) => any
+
+interface LabeledInputProps {
+  strongLabel?: string
+  patternConfig?: PatternConfig
+  label: string
+  register: Register
+  required: boolean
+  name: string
+  errors: Errors
+  defaultValue?: string
+}
+
+const getError = (errors: Errors, name: string): string | undefined => {
+  let result
+  if (Object.prototype.hasOwnProperty.call(errors, name)) {
+    const r = errors[name]
+    result = r.type
+  }
+  return result
+}
+
+const isError = (errors: Errors, name: string): boolean => getError(errors, name) !== undefined
+
+export function LabeledInput (props: LabeledInputProps): ReactElement {
+  const { strongLabel, label, register, required, patternConfig = {}, name, errors, defaultValue } = props
+  let helperText: string | undefined
+  // fix error where pattern wouldn't match if input wasn't filled out even if required was set to false
+  if (!required) {
+    patternConfig.regexp = /.?/
+  }
+  if (errors[name]?.type === 'required') {
+    helperText = 'Input is required'
+  } else if (errors[name]?.type === 'pattern') {
+    // Must have a regex pattern description so users can see what's going wrong *
+    helperText = patternConfig.description
+  }
+
+  /* default regex pattern is to accept any input. Otherwise, use input pattern */
+  const textField = (
+    <TextField
+      error={isError(errors, name)}
+      helperText={helperText}
+      fullWidth={patternConfig.mask === undefined}
+      defaultValue={defaultValue}
+      inputRef={register({ required: required, pattern: patternConfig.regexp })}
+      name={name}
+      variant="filled"
+    />
+  )
+
+  /* if there is a mask prop, create masked textfield rather than standard */
+  let input = textField
+  if (patternConfig.mask !== undefined) {
+    input = (
+      <InputMask
+        mask={patternConfig.mask}
+        alwaysShowMask={true}
+        defaultValue={defaultValue}
+      >
+        {textField}
+      </InputMask>
+    )
+  }
+
+  return (
+    <div>
+      <Box display="flex" justifyContent="flex-start">
+        <p><strong>{strongLabel}</strong>{label}</p>
+      </Box>
+      <Box display="flex" justifyContent="flex-start">
+        {input}
+      </Box>
+    </div>
+  )
+}
+
+interface BaseDropdownProps {
```

Comment:
Can other interfaces like this and LabeledDropdownProps<A,> and LabeledInputProps also be factored into a separate interfaces file

Reply:
Yeah this file is doing way too much, I'll break it up.